examples/audio/transformer_asr.py

```diff
@@ -0,0 +1,523 @@
+    stddevs = tf.math.reduce_std(x, 1, keepdims=True)
+    x = tf.divide(tf.subtract(x, means), stddevs)
+    audio_len = tf.shape(x)[0]
+    # padding to 10 seconds
+    pad_len = 2754
+    paddings = tf.constant([[0, pad_len], [0, 0]])
+    x = tf.pad(x, paddings, "CONSTANT")[:pad_len, :]
+    return x
+
+
+def create_audio_ds(data):
+    flist = [_["audio"] for _ in data]
+    audio_ds = tf.data.Dataset.from_tensor_slices(flist)
+    audio_ds = audio_ds.map(
+        path_to_audio, num_parallel_calls=tf.data.experimental.AUTOTUNE
+    )
+    return audio_ds
+
+
+def create_tf_dataset(data, bs=4):
+    audio_ds = create_audio_ds(data)
+    text_ds = create_text_ds(data)
+    ds = tf.data.Dataset.zip((audio_ds, text_ds))
+    ds = ds.map(lambda x, y: {"src": x, "trg": y})
+    ds = ds.batch(bs)
+    ds = ds.prefetch(tf.data.experimental.AUTOTUNE)
+    return ds
+
+
+"""
+Check contents of one batch
+"""
+
+
+split = int(len(data) * 0.99)
+train_data = data[:split]
+test_data = data[split:]
+
+ds = create_tf_dataset(train_data, bs=64)
+val_ds = create_tf_dataset(test_data, bs=4)
+
+for i in ds.take(1):
+    print(i["src"].shape)
+    print(i["trg"])
+
+
+"""
+## Callbacks to display predictions and to change learning rate
+"""
+
+
+class DisplayOutputs(keras.callbacks.Callback):
+    def __init__(
+        self, batch, idx_to_token, trg_start_token_idx=27, trg_end_token_idx=28
+    ):
+        """ Displays a batch of outputs after every epoch 
+        Arguments:
+        - batch: test batch containing the keys "src" and "trg"
+        - idx_to_token: a List containing the vocabulary tokens corresponding to their indices
+        - trg_start_token_idx: start token index in the target vocabulary
+        - trg_end_token_idx: end token index in the target vocabulary
+        """
+        self.batch = batch
+        self.trg_start_token_idx = trg_start_token_idx
+        self.trg_end_token_idx = trg_end_token_idx
+        self.idx_to_char = idx_to_token
+
+    def on_epoch_end(self, epoch, logs=None):
+        if epoch % 5 != 0:
+            return
+        src = self.batch["src"]
+        trg = self.batch["trg"].numpy()
+        bs = tf.shape(src)[0]
+        preds = self.model.generate(src, self.trg_start_token_idx)
+        preds = preds.numpy()
+        for i in range(bs):
+            target = ""
+            for idx in trg[i, :]:
+                target += "" + self.idx_to_char[idx]
+            prediction = ""
+            over = False
+            for idx in preds[i, :]:
+                if over:  # Add padding token once end token has beeen predicted
+                    prediction += "-"
+                    continue
+                if idx == self.trg_end_token_idx:
+                    over = True
+                prediction += "" + self.idx_to_char[idx]
+            print(f"target:     {target}")
+            print(f"prediction: {prediction}")
+            print()
+
+
+def scheduler(epoch, lr):
```

Comment:
It would be simpler to use one of the built-in learning rate schedules, like this one: https://keras.io/api/optimizers/learning_rate_schedules/inverse_time_decay/

Reply:
I could not find a built-in schedule for warmup. https://keras.io/api/optimizers/learning_rate_schedules/
A simple decay might lead to a performance drop or instability. I've observed warmups to be particularly important with transformers.

Reply:
You could structure your own schedule as a learning rate schedule -- just a callable class where the `__call__` takes a `step` arg.

See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/optimizer_v2/learning_rate_schedule.py